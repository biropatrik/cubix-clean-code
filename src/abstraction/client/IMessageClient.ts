export interface IMessageClient {
    sendNotification(message: string): void;
}