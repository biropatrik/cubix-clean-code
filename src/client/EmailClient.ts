import { IMessageClient } from "../abstraction/client/IMessageClient";

export class EmailClient implements IMessageClient {

    readonly EMAIL: string = 'Email: ';

    sendNotification(message: string): void {
        console.log(this.EMAIL.concat(message));
    }

}