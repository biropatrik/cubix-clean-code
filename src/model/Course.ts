import { StudentAlreadyAddedException } from "../exception/StudentAlreadyAddedException";
import { Student } from "./Student";

export class Course {

    constructor(
        private name: string,
        private students: Student[] = []) { }

    public getName() {
        return this.name;
    }

    public addStudent(student: Student) {
        if (this.students.includes(student)) {
            throw new StudentAlreadyAddedException('Student already added to the course!');
        }
        this.students.push(student);
    }
}