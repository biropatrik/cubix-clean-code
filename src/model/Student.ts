import { Course } from "./Course";
import { Person } from "./Person";

export class Student extends Person {

    constructor(name: string, courses?: Course[]) {
        super(name, courses);
    }
}