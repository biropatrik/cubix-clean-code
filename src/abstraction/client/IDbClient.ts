import { Course } from "../../model/Course";
import { Student } from "../../model/Student";

export interface IDbClient {
    getCourseByName(name: string): Course | undefined;
    getAllCourses(): Course[];
    addCourse(course: Course): void;
    addStudentToCourse(student: Student, courseName: string): void;
}