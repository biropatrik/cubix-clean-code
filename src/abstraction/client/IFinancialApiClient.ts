import { Student } from "../../model/Student";

export interface IFinancialApiClient {
    getIsOrderPayed(student: Student): Promise<boolean>;
    payForCourse(student: Student, courseName: string): Promise<void>;
}