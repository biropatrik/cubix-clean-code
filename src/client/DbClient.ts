import { IDbClient } from "../abstraction/client/IDbClient";
import { Course } from "../model/Course";
import { Student } from "../model/Student";

export class DbClient implements IDbClient {

    private courses: Course[] = [new Course('Data Analytics'), new Course('Project Management'), new Course('Java')]

    getCourseByName(name: string): Course | undefined {
        let result = undefined;
        for (const course of this.courses) {
            if (course.getName() === name) {
                result = course;
            }
        }
        return result;
    }

    getAllCourses(): Course[] {
        return this.courses;
    }

    addCourse(course: Course): void {
        this.courses.push(course);
    }

    addStudentToCourse(student: Student, courseName: string): void {
        for (const course of this.courses) {
            if (course.getName() === courseName) {
                course.addStudent(student);
            }
        }
    }
}