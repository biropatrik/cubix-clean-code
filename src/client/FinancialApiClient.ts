import { IFinancialApiClient } from "../abstraction/client/IFinancialApiClient";
import { IFinancialService } from "../abstraction/service/IFinancialService";
import { NetworkException } from "../exception/NetworkException";
import { UnknownException } from "../exception/UnknownException";
import { Student } from "../model/Student";

export class FinancialApiClient implements IFinancialApiClient {

    constructor(private financialService: IFinancialService) { }

    async getIsOrderPayed(student: Student): Promise<boolean> {
        try {
            return await this.financialService.getIsOrderPayed(student);
        } catch (error) {
            if (error instanceof NetworkException) {
                throw error;
            }
            throw new UnknownException('Unknown error happened.');
        }
    }

    async payForCourse(student: Student, courseName: string): Promise<void> {
        try {
            await this.financialService.payForCourse(student, courseName);
        } catch (error) {
            if (error instanceof NetworkException) {
                throw error;
            }
            throw new UnknownException('Unknown error happened.');
        }
    }
}