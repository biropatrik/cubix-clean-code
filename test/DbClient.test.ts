import { DbClient } from "../src/client/DbClient"
import { Course } from "../src/model/Course";
import { Student } from "../src/model/Student";

describe('DbClient tests', () => {
    let dbClient: DbClient;

    beforeEach(() => {
        dbClient = new DbClient();
    })

    it('should return a course by name', () => {
        // Arrange
        const courseName = 'Java';

        // Act
        const result = dbClient.getCourseByName(courseName);

        // Assert
        expect(result).toMatchSnapshot();
    })

    it('should return all courses', () => {
        // Act
        const result = dbClient.getAllCourses();

        // Assert
        expect(result).toMatchSnapshot();
        expect(result).toHaveLength(3);
    })

    it('should add new course', () => {
        // Arrange
        const newCourse = new Course('Power BI');

        // Act
        dbClient.addCourse(newCourse);
        const result = dbClient.getAllCourses();

        // Assert
        expect(result).toMatchSnapshot();
        expect(result).toHaveLength(4);
    })

    it('should add student to a course', () => {
        // Arrange
        const courseName = 'Java';
        const student = new Student('John');
        const expectedCourse = new Course(courseName, [student]);

        // Act
        dbClient.addStudentToCourse(student, courseName);
        const result = dbClient.getCourseByName(courseName);

        // Assert
        expect(result).toEqual(expectedCourse);
    })
})