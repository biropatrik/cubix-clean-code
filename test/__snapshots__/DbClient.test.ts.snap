// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DbClient tests should add new course 1`] = `
[
  Course {
    "name": "Data Analytics",
    "students": [],
  },
  Course {
    "name": "Project Management",
    "students": [],
  },
  Course {
    "name": "Java",
    "students": [],
  },
  Course {
    "name": "Power BI",
    "students": [],
  },
]
`;

exports[`DbClient tests should return a course by name 1`] = `
Course {
  "name": "Java",
  "students": [],
}
`;

exports[`DbClient tests should return all courses 1`] = `
[
  Course {
    "name": "Data Analytics",
    "students": [],
  },
  Course {
    "name": "Project Management",
    "students": [],
  },
  Course {
    "name": "Java",
    "students": [],
  },
]
`;
